# Task 4.9 Part 1

This script contains the following points:

    01. Improting Libraries
    02. Importing CSV/PKL files
    03. Work for Task Part 1
    04. Exporting Pickle format    

## 01. Importing Libraries

In [4]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## 02. Importing CSV/PKL files

In [5]:
# Identifying pathway for import for ease with multiple imports
path = r'C:\Users\Chris Arnold\Documents\Instacart Basket Analysis'

In [6]:
# Importing customers.csv
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

## 03. Work for Task Part 1

### Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

Looking for columns to rename/drop

In [9]:
# Looking at the head ofthe dataframe
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

There doesn't seem to be any columns that need to be removed from this data set, as they all could be useful bits for an analysis. We do not know the questions that are going to be asked, so removing data would be foolish at this point. There are some columns that can be renamed though.

In [10]:
# Renaming columns
customers.rename(columns = {'First Name' : 'First_Name'}, inplace = True)

In [11]:
# Renaming columns
customers.rename(columns = {'Surnam' : 'Last_Name'}, inplace = True)

In [12]:
# Renaming columns
customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [13]:
# Renaming columns
customers.rename(columns = {'user_id' : 'User_ID'}, inplace = True)

In [14]:
# Renaming columns
customers.rename(columns = {'date_joined' : 'Date_Joined'}, inplace = True)

In [15]:
# Renaming columns
customers.rename(columns = {'n_dependants' : 'Number_of_Dependants'}, inplace = True)

In [16]:
# Renaming columns
customers.rename(columns = {'fam_status' : 'Family_Status'}, inplace = True)

In [17]:
# Renaming columns
customers.rename(columns = {'income' : 'Income'}, inplace = True)

In [18]:
# Checking work
customers.head()

User_ID First_Name Last_Name  Gender       State  Age Date_Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number_of_Dependants Family_Status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

Column names look good!

Checking data types

In [19]:
#checking data types
customers['User_ID'].dtype

dtype('int64')

In [20]:
# Changing User_ID to a string because it is just a different representation of the customer name (even if random or sequential)
customers['User_ID'] = customers['User_ID'].astype('str')

In [21]:
customers['User_ID'].dtype

dtype('O')

In [22]:
#checking data types
customers['First_Name'].dtype

dtype('O')

In [24]:
#checking data types
customers['Last_Name'].dtype

dtype('O')

In [25]:
#checking data types
customers['Gender'].dtype

dtype('O')

In [26]:
#checking data types
customers['State'].dtype

dtype('O')

In [27]:
#checking data types
customers['Age'].dtype

dtype('int64')

In [28]:
#checking data types
customers['Date_Joined'].dtype

dtype('O')

In [29]:
#checking data types
customers['Number_of_Dependants'].dtype

dtype('int64')

In [30]:
#checking data types
customers['Family_Status'].dtype

dtype('O')

In [31]:
#checking data types
customers['Income'].dtype

dtype('int64')

Data types now all check out!

### Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [32]:
# Checking descriptive statistics on the data set
customers.describe()

Age  Number_of_Dependants         Income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

The numbers here all look like resonable ranges, means, standard deviations, min values, max values, and the counts are all the same.

In [33]:
# Checking for null/missing values
customers.isnull().sum()

User_ID                     0
First_Name              11259
Last_Name                   0
Gender                      0
State                       0
Age                         0
Date_Joined                 0
Number_of_Dependants        0
Family_Status               0
Income                      0
dtype: int64

There are 11,259 missing first names in this data set. This isn't horribly concerning because we have the rest of the information filled in for everything else. At this point I will leave them missing assuming that some people just give a family name and do not want to give their first name.

In [34]:
# Looking for duplicates
customers_dups = customers[customers.duplicated()]

In [35]:
customers_dups.head()

Empty DataFrame
Columns: [User_ID, First_Name, Last_Name, Gender, State, Age, Date_Joined, Number_of_Dependants, Family_Status, Income]
Index: []

There are no duplicates in this data set

Please also note that the mixed data types were already dealt with easrlier and all columns show uniform data type.

### Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type! Hint: Make sure the key columns are the same data type!

In [37]:
# Importing the combined data set from task 4.8
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4.8.pkl'))

In [38]:
# Checking the head of the ords_prods_merge to find the key column
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average Orders         10   
1  Regularly busy    Slowest days        Average Orders         10   
2  Regularly busy    Slowest days           Most Orders         10   
3      Least busy  Regularly busy        Average Orders         10   
4      Least busy  Regularly busy           Most Orders         10   

   loyalty_flag  Total_spent Mean_prices Spender_type  \
0  New customer         14.0    6.367797  Low_spender   
1  New customer         14.0    6.367797  Low_spender   
2  New customer         14.0    6.367797  Low_spender   
3  New customer         14.0    6.367797  Low_spender   
4  New customer         14.0    6.367797  Low_spender   

  Median_days_since_prior_order     Customer_frequency  
0                          20.5  Non_frequent_customer  
1                          20.5  Non_frequent_customer  
2                          20.5  Non_frequent_customer  
3                          20.5  Non_frequent_customer  
4                          20.5  Non_frequent_customer  

[5 rows x 28 columns]

In [39]:
# checking head of customers dataframe
customers.head()

User_ID First_Name Last_Name  Gender       State  Age Date_Joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number_of_Dependants Family_Status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

It looks like we will be able to pivot around the user_id column very easily.

In [40]:
# Changing column on ords_prods_merge to match customers USER_ID
# Renaming columns
ords_prods_merge.rename(columns = {'user_id' : 'User_ID'}, inplace = True)

In [41]:
# Checking work
ords_prods_merge.head()

order_id  User_ID eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average Orders         10   
1  Regularly busy    Slowest days        Average Orders         10   
2  Regularly busy    Slowest days           Most Orders         10   
3      Least busy  Regularly busy        Average Orders         10   
4      Least busy  Regularly busy           Most Orders         10   

   loyalty_flag  Total_spent Mean_prices Spender_type  \
0  New customer         14.0    6.367797  Low_spender   
1  New customer         14.0    6.367797  Low_spender   
2  New customer         14.0    6.367797  Low_spender   
3  New customer         14.0    6.367797  Low_spender   
4  New customer         14.0    6.367797  Low_spender   

  Median_days_since_prior_order     Customer_frequency  
0                          20.5  Non_frequent_customer  
1                          20.5  Non_frequent_customer  
2                          20.5  Non_frequent_customer  
3                          20.5  Non_frequent_customer  
4                          20.5  Non_frequent_customer  

[5 rows x 28 columns]

Now that they match we can proceed with the merge a full outer join seems appropriate since we want all the data from both without duplication of the User_ID column

Making sure that the columns are both the same data type

In [45]:
#checking data types
customers['User_ID'].dtype

dtype('O')

In [46]:
ords_prods_merge['User_ID'].dtype

dtype('int64')

In [47]:
# Changing User_ID to a string because it is just a different representation of the customer name (even if random or sequential)
ords_prods_merge['User_ID'] = ords_prods_merge['User_ID'].astype('str')

In [48]:
#Checking
ords_prods_merge['User_ID'].dtype

dtype('O')

In [52]:
# Performing the merge
ords_prods_customers_merged = customers.merge(ords_prods_merge, on = 'User_ID')

In [55]:
# Checking result
ords_prods_customers_merged.head(10)

User_ID First_Name Last_Name  Gender     State  Age Date_Joined  \
0   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
5   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
6   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
7   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
8   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
9   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   Number_of_Dependants Family_Status  Income  ...     busiest_day  \
0                     3       married  165665  ...  Regularly busy   
1                     3       married  165665  ...  Regularly busy   
2                     3       married  165665  ...  Regularly busy   
3                     3       married  165665  ...  Regularly busy   
4                     3       married  165665  ...  Regularly busy   
5                     3       married  165665  ...  Regularly busy   
6                     3       married  165665  ...  Regularly busy   
7                     3       married  165665  ...  Regularly busy   
8                     3       married  165665  ...  Regularly busy   
9                     3       married  165665  ...  Regularly busy   

     busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
0  Regularly busy         Average Orders          8  New customer   
1  Regularly busy            Most Orders          8  New customer   
2  Regularly busy            Most Orders          8  New customer   
3  Regularly busy         Average Orders          8  New customer   
4    Slowest days            Most Orders          8  New customer   
5  Regularly busy            Most Orders          8  New customer   
6  Regularly busy         Average Orders          8  New customer   
7  Regularly busy            Most Orders          8  New customer   
8  Regularly busy            Most Orders          8  New customer   
9  Regularly busy         Average Orders          8  New customer   

   Total_spent  Mean_prices  Spender_type  Median_days_since_prior_order  \
0         12.6     7.988889   Low_spender                           19.0   
1         12.6     7.988889   Low_spender                           19.0   
2         12.6     7.988889   Low_spender                           19.0   
3         12.6     7.988889   Low_spender                           19.0   
4         12.6     7.988889   Low_spender                           19.0   
5         12.6     7.988889   Low_spender                           19.0   
6         12.6     7.988889   Low_spender                           19.0   
7         12.6     7.988889   Low_spender                           19.0   
8         12.6     7.988889   Low_spender                           19.0   
9         12.6     7.988889   Low_spender                           19.0   

  Customer_frequency  
0   Regular_customer  
1   Regular_customer  
2   Regular_customer  
3   Regular_customer  
4   Regular_customer  
5   Regular_customer  
6   Regular_customer  
7   Regular_customer  
8   Regular_customer  
9   Regular_customer  

[10 rows x 37 columns]

In [56]:
# Checking for what exactly happened
ords_prods_customers_merged['_merge'].value_counts()

both          32406041
left_only            0
right_only           0
Name: _merge, dtype: int64

In [57]:
ords_prods_customers_merged.isnull().sum()

User_ID                                 0
First_Name                        1775189
Last_Name                               0
Gender                                  0
State                                   0
Age                                     0
Date_Joined                             0
Number_of_Dependants                    0
Family_Status                           0
Income                                  0
order_id                                0
eval_set                                0
order_number                            0
order_dow                               0
order_hour_of_day                       0
days_since_prior_order            2076201
product_id                              0
add_to_cart_order                       0
reordered                               0
_merge                                  0
Unnamed: 0                              0
product_name                            0
aisle_id                                0
department_id                     

Cleaning up un needed columns, Noticed the prive_range_loc is an un needed accidental duplicate created column with no needed data.

In [60]:
ords_prods_customers_merged = ords_prods_customers_merged.drop(columns = ['prive_range_loc'])

In [61]:
# Checking work
ords_prods_customers_merged.isnull().sum()

User_ID                                0
First_Name                       1775189
Last_Name                              0
Gender                                 0
State                                  0
Age                                    0
Date_Joined                            0
Number_of_Dependants                   0
Family_Status                          0
Income                                 0
order_id                               0
eval_set                               0
order_number                           0
order_dow                              0
order_hour_of_day                      0
days_since_prior_order           2076201
product_id                             0
add_to_cart_order                      0
reordered                              0
_merge                                 0
Unnamed: 0                             0
product_name                           0
aisle_id                               0
department_id                          0
prices          

In [62]:
# Double checking for duplicates
ords_prods_customers_merged_dups = ords_prods_customers_merged[ords_prods_customers_merged.duplicated()]

In [63]:
ords_prods_customers_merged_dups.head()

Empty DataFrame
Columns: [User_ID, First_Name, Last_Name, Gender, State, Age, Date_Joined, Number_of_Dependants, Family_Status, Income, order_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered, _merge, Unnamed: 0, product_name, aisle_id, department_id, prices, price_range_loc, busiest_day, busiest_days, busiest_period_of_day, max_order, loyalty_flag, Total_spent, Mean_prices, Spender_type, Median_days_since_prior_order, Customer_frequency]
Index: []

[0 rows x 36 columns]

No duplicates Found!

## 04. Exporting to Pickle format

In [64]:
# Export data to pkl

ords_prods_customers_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_merged.pkl'))